apps/els_lsp/src/els_fungraph.erl

```diff
@@ -0,0 +1,38 @@
+-module(els_fungraph).
+
+-type gnode() :: #{id := _ID, _ => _}.
+-type edges_fun() :: fun((gnode()) -> list(gnode())).
+-opaque graph() :: {?MODULE, edges_fun()}.
+
+-export_type([gnode/0]).
+-export_type([graph/0]).
+
+-export([new/1]).
+-export([traverse/4]).
+
+-spec new(edges_fun()) -> graph().
+new(EdgesFun) ->
+    {?MODULE, EdgesFun}.
+
+-type acc_fun(AccT) :: fun((gnode(), AccT) -> AccT).
+
+-spec traverse(acc_fun(AccT), AccT, _From :: gnode(), graph()) -> AccT.
+traverse(AccFun, Acc, From, G) ->
+  traverse(AccFun, Acc, [From], sets:new(), G).
+
+-spec traverse(acc_fun(AccT), AccT, _From :: gnode(), sets:set(_NodeID), graph()) -> AccT.
+traverse(AccFun, Acc, Nodes, Visited, G = {?MODULE, EdgesFun}) ->
+  AdjacentNodes = [
```

Comment:
do I see correctly that `AdjacentNodes` might have duplicates if multiple Nodes reference the same document? (so an usort could help to avoid duplicates in `AccNext`)

Reply:
Again, spot on! I fixed it by rewriting into a depth-first traversal, for a bit of extra efficiency (hopefully).